lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,642 @@
+    one_block_return : FixedPoint, optional
+        An estimate of the variable return expected for the next block, by default None.
+    """
+    # pylint: disable=too-many-arguments
+    if one_block_return is None:
+        one_block_return = FixedPoint(1)
+    k_t = calc_k(
+        share_price,
+        initial_share_price,
+        share_reserves,
+        bond_reserves,
+        time_stretch,
+    )
+    y_term = (bond_reserves + bonds_in) ** (fp1 - time_stretch)
+    z_val = (k_t - y_term) / (share_price / initial_share_price)
+    z_val = z_val ** (fp1 / (fp1 - time_stretch))
+    z_val /= initial_share_price
+    # z_val *= one_block_return
+    spot_price = calc_spot_price_local(initial_share_price, share_reserves, fp0, bond_reserves, time_stretch)
+    price_discount = fp1 - spot_price
+    amount_in_shares = max(fp0, share_reserves - z_val)
+    curve_fee_rate = price_discount * curve_fee
+    curve_fee_amount_in_shares = amount_in_shares * curve_fee_rate
+    gov_fee_amount_in_shares = curve_fee_amount_in_shares * gov_fee
+    # applying fee means you get LESS shares out for the same amount of bonds IN
+    amount_to_user_in_shares = amount_in_shares - curve_fee_amount_in_shares
+    return amount_to_user_in_shares, curve_fee_amount_in_shares, gov_fee_amount_in_shares
+
+
+def calc_reserves_to_hit_target_rate(
+    target_rate: FixedPoint, interface: HyperdriveInterface
+) -> tuple[FixedPoint, FixedPoint]:
+    """Calculate bonds needed to hit target rate.
+
+    Arguments
+    ---------
+    target_rate : Decimal
+        The target rate.
+    interface : HyperdriveInterface
+        The Hyperdrive API interface object.
+
+    Returns
+    -------
+    tuple[FixedPoint, FixedPoint] containing:
+        total_shares_needed : FixedPoint
+            Total amount of shares needed to be added into the pool to hit the target rate.
+        total_bonds_needed : FixedPoint
+            Total amount of bonds needed to be added into the pool to hit the target rate.
+    """
+    # variables
+    predicted_rate = fp0
+    pool_config = interface.pool_config.copy()
+    pool_info = interface.pool_info.copy()
+
+    variable_rate = interface.variable_rate
+    one_block_return = (fp1 + variable_rate) ** (
+        fp12 / fp_seconds_in_year
+    )  # attempt to see if this improves accuracy, it's 1e-8 difference
+
+    iteration = 0
+    start_time = time.time()
+    total_shares_needed = fp0
+    total_bonds_needed = fp0
+    print(f"Targetting {float(target_rate):.2%} from {float(interface.fixed_rate):.2%}")
+    while float(abs(predicted_rate - target_rate)) > tolerance:
+        iteration += 1
+        target_bonds = calc_bond_reserves(
+            pool_info["shareReserves"],
+            pool_config["initialSharePrice"],
+            target_rate,
+            pool_config["positionDuration"],
+            pool_config["invTimeStretch"],
+        )
+        bonds_needed = (target_bonds - pool_info["bondReserves"]) / fp2
```

Comment:
why is this divided by 2?

Reply:
to get halfway there 🤪
but because if  you just added the total amount of bonds needed, that ignores the fact that you're also extracting an equal value of shares at the same time. so to actually hit the desired ratio (ignoring nonlinearity) you have to put in half that bond amount, while taking out half that value in share reserves, hitting the desired ratio.

Reply:
ahh ok, I got it. So this puts a bounds on the accuracy right? it would be more accurate if we divided by a larger number?

Reply:
there might be another factor we could eyeball to improve this guess, but I can't think of it.
the obvious one is just going halfway, since that takes into account what happens on the other side of the x/y relationship.